### ISSUE TITLE
AppCrashes in BTPayPalDriver on iOS 9.2, iPhone 6s, Braintree pod 4.1.3

### ISSUE THREAD

**Issue by nikitabelosludtcev** [braintree/braintree_ios#218]
Application crashes when start in BTPayPalDriver on + (void)load  method.

 if (self == [BTPayPalDriver class]) {
     PayPalClass = [PayPalOneTouchCore class]; <------  crashes here 

Terminating app due to uncaught exception 'NSInvalidUnarchiveOperationException', reason: ' -[NSKeyedUnarchiver decodeObjectForKey:]: cannot decode object of class (PPOTConfiguration) for key (root); the class may be defined in source code or a library that is not linked'
*** First throw call stack:
(0x182059900 0x1816c7f80 0x1829c0f4c 0x1829c1104 0x18298f1c8 0x18298ef90 0x18298e124 0x10040bf78 0x1003faf88 0x1816c8fc8 0x1816cf8f8 0x1816d9db8 0x1816c8fc8 0x1816cf8f8 0x1816d9db8 0x10024bae8 0x1816ca27c 0x1816cae30 0x1200cdd3c 0x1200dc960 0x1200dbb20 0x1200dbbd4 0x1200ce08c 0x1200d20fc 0x1200cd044)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb)

**Comment by demerino** [comment]
What does your Podfile look like? And does the error still occur if you update to 4.2.1?

**Comment by nikitabelosludtcev** [comment]
Hello, thanks for reply.

**Comment by skunkworks** [comment]
If you're on < `4.2` of the Braintree iOS SDK, the PayPal One Touch static library is not being linked by your app. Make sure that `libPayPalOneTouchCore.a` is linked under Project > Build Phases > Link Binary With Libraries.

**Comment by nikitabelosludtcev** [comment]
Hello, thanks for reply.
My podfile is:

```
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '8.0'

pod 'AFNetworking', '~> 2.6.3'
pod 'Reachability'
pod 'SDWebImage'
pod 'SVProgressHUD'
pod 'MHVideoPhotoGallery'
pod 'RSKImageCropper'
pod 'StreamingKit', '~> 0.1'
pod 'UIActivityIndicator-for-SDWebImage'
pod 'ECSlidingViewController', '~> 2.0.3'
pod 'Mantle', '~> 2.0'
pod 'SVPullToRefresh'
pod 'DZNPhotoPickerController/Editor'
pod 'IQAudioRecorderController'
pod 'JSQMessagesViewController'
pod 'CWStatusBarNotification', '~> 2.1.0'
pod 'Fabric'
pod 'Crashlytics'
pod 'UITextView+Placeholder'
pod 'DLImageLoader', '~> 2.1'
pod 'KWKeyboardListener', '~> 1.0'
pod 'Braintree'
pod 'CardIO'
pod 'QRCodeReaderViewController', '~> 4.0.1'
pod "AHKNavigationController"
```
 
I use Braintree (4.1.3)

**Comment by nikitabelosludtcev** [comment]
> Make sure that libPayPalOneTouchCore.a is linked under Project > Build Phases > Link Binary With Libraries.

Ok I will do. But it's strange that application crashed only on one device. On other devices app works Good.

**Comment by skunkworks** [comment]
Also, it is highly likely that updating to version 4.2.1 will eliminate this error. The PayPal OTC library is now part of our code repository instead of being a vendored static library.

**Comment by Japspre** [comment]
This happens when you would have upgraded braintree version to 4.2.1 f.e. and then downgraded 4.1.3. On simulator you can reset the simulator or delete the app, but on device there's some key which is persisted in the keychain which will cause the crash. It just shows that braintree has poor backwards compatibility and luckily this didn't happen in our production app. 

If you wish to stay on v4.1.3 you can manually fix this by commenting the `+ (void)load` method in `BTPayPalDriver.m`. Then using the following block of code will clear your secure keychain:

> void (^deleteAllKeysForSecClass)(CFTypeRef) = ^(CFTypeRef secClass) {
    id dict = @{(__bridge id)kSecClass: (__bridge id)secClass};
    OSStatus result = SecItemDelete((__bridge CFDictionaryRef) dict);
    NSAssert(result == noErr || result == errSecItemNotFound, @"Error deleting keychain data (%d)", (int)result);
};
deleteAllKeysForSecClass(kSecClassGenericPassword);
deleteAllKeysForSecClass(kSecClassInternetPassword);
deleteAllKeysForSecClass(kSecClassCertificate);
deleteAllKeysForSecClass(kSecClassKey);
deleteAllKeysForSecClass(kSecClassIdentity);

**Comment by nikitabelosludtcev** [comment]
@Japspre
>  but on device there's some key which is persisted in the keychain which will cause the crash

If I am not mistaken key chain is removed when deleting application. Why it's still storing info in keychain?

**Comment by Japspre** [comment]
@nikitabelosludtcev you are mistaking NSUserDefaults with the keyChain. 
http://stackoverflow.com/questions/12090136/difference-between-keychain-and-nsuserdefault
http://stackoverflow.com/questions/18911434/will-items-in-ios-keychain-survive-app-uninstall-and-reinstall

**Comment by skunkworks** [comment]
We have duplicated this issue here, a fix is incoming shortly. Thanks for your patience.

**Comment by skunkworks** [comment]
Hi, we've released 4.2.2, which should fix this issue moving forward.
